### PR TITLE
Stop perpetually writing `mb_artistid`, `mb_albumartistid` and `albumtypes` fields

### LINKED ISSUE #5265: `beet write` continuously writes mb_artistid and mb_albumartistid (1.7.0_pre20240501)
When I execute the `beet write` command it will try to write **mb_artistid** and **mb_albumartistid** tags every single time the program is executed.

It appeared when version **1.7.0_pre20240501** was installed on my Gentoo system. I got this version because I configured portage, the package manager, to install the "unstable" version (`media-sound/beets ~amd64`).

Previous version was 1.6.0 and it didn't show this behavior.

I will probably downgrade to 1.6.0, but I wanted to submit this bug in case there is something specific with the 1.7.0 version.

### Problem

Running this command in verbose (`-vv`) mode:

```sh
$ beet -vv write album:"For All Kings"
```

Led to this problem:

```
$ beet -vv write album:"For All Kings" 
overlaying configuration: /home/mcarpentier/Music/.beets/config.yaml
no user configuration found at /home/mcarpentier/.config/beets/config.yaml
data directory: /home/mcarpentier/.config/beets
plugin paths: 
lyrics: Disabling google source: no API key configured.
fetchart: google: Disabling art source due to missing key
fetchart: lastfm: Disabling art source due to missing key
Sending event: pluginload
library database: /home/mcarpentier/Music/.beets/BeetMusicLibrary.blb
library directory: /home/mcarpentier/Music
Sending event: library_opened
Parsed query: AndQuery([SubstringQuery('album', 'For All Kings', fast=True)])
Parsed sort: NullSort()
Anthrax - For All Kings - Impaled
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - You Gotta Believe
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Monster at the End
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - For All Kings
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Breathing Lightning
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Breathing Out
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Suzerain
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Evil Twin
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Blood Eagle Wings
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Defend Avenge
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - All of Them Thieves
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - This Battle Chose Us
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Anthrax - For All Kings - Zero Tolerance
  mb_artistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
  mb_albumartistid:  -> b616d6f0-ec1f-4c69-8a79-12a97ece7372
Sending event: write
Sending event: after_write
Sending event: database_change
Sending event: cli_exit
```

I filtered on a specific album because I don't want tags to be overwritten on the whole library.  I only have FLAC file in my library.

It will repeat every time I run the `write` command. 

The files are modified according to the "stat" output (`Modify` matches when I ran the `beet write` command).

```
$ stat 01\ Impaled.flac 
  File: 01 Impaled.flac
  Size: 11241570  	Blocks: 21960      IO Block: 4096   regular file
Device: 0,36	Inode: 13139015    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/mcarpentier)   Gid: ( 1000/mcarpentier)
Access: 2022-10-23 19:39:04.298317103 +0200
Modify: 2024-05-27 18:56:38.947141933 +0200
Change: 2024-05-27 18:56:38.947141933 +0200
 Birth: 2022-10-23 19:39:04.298317103 +0200
```

Same thing happens if I disable plugins (`beet -vv --plugins= write album:"For All Kings"`)

The files are changed because they will be included in the next backup.

### Setup

* OS: Gentoo
* Python version: 3.11.9
* beets version: 1.7.0_pre20240501
* Turning off plugins made problem go away (yes/no): 

My configuration (output of `beet config`) is:

```yaml
lyrics:
    bing_lang_from: []
    auto: yes
    force: yes
    fallback: ''
    bing_client_secret: REDACTED
    bing_lang_to:
    google_API_key: REDACTED
    google_engine_ID: REDACTED
    genius_api_key: REDACTED
    local: no
    synced: no
    sources:
    - google
    - genius
    - tekstowo
    - lrclib
    dist_thresh: 0.1
# --------------- Main ---------------

library: /home/mcarpentier/Music/.beets/BeetMusicLibrary.blb
directory: /home/mcarpentier/Music

ui:
    color: yes
    colors:
        text_success: green
        text_warning: yellow
        text_error: red
        text_highlight: red
        text_highlight_minor: lightgray
        action_default: turquoise
        action: blue

import:
    copy: no
    write: yes
    move: yes

# --------------- Plugins ---------------

plugins: embedart fetchart info lyrics edit
embedart:
    auto: yes
    maxwidth: 0
    compare_threshold: 0
    ifempty: no
    remove_art_file: no
    quality: 0
paths:
    default: $albumartist/$album%aunique{}/$track $title
    singleton: Non-Album/$artist/$title
    comp: Compilations/$album%aunique{}/$track $title
fetchart:
    auto: yes
    minwidth: 0
    maxwidth: 0
    quality: 0
    max_filesize: 0
    enforce_ratio: no
    cautious: no
    cover_names:
    - cover
    - front
    - art
    - album
    - folder
    sources:
    - filesystem
    - coverart
    - itunes
    - amazon
    - albumart
    - cover_art_url
    store_source: no
    high_resolution: no
    deinterlace: no
    cover_format:
    google_key: REDACTED
    google_engine: 001442825323518660753:hrh5ch1gjzm
    fanarttv_key: REDACTED
    lastfm_key: REDACTED
edit:
    albumfields: album albumartist
    itemfields: track title artist album
    ignore_fields: id path
```

### PR DESCRIPTION
**Pull request by snejus**
This PR fixes an issue where the `beet write` command repeatedly shows differences for certain fields (`mb_artistid`, `mb_albumartistid`, `albumtype`) even after writing the tags. 
This happens because these fields are actually stored as lists in the media files (`mb_artistids`, `mb_albumartistids`, `albumtypes`), but beets maintains both single and list versions in its database.

This PR addresses this issue in a non-invasive way: the fix ensures consistency between single fields and their list counterparts by:
1. When setting a single field value, making it the first element of the corresponding list
2. When setting a list, using its first element as the single field value

This resolves long-standing issues #5265, #5371, and #4715 where users experienced persistent "differences" in these fields despite writing tags multiple times.

Changes:
- Added `ensure_consistent_list_fields()` function to synchronize field pairs
- Applied this function during metadata application
- Added tests for all field combinations

Fixes #5265, #5371, #4715

**Note**: your music needs to be reimported with `beet import -LI` or synchronised with `beet mbsync` in order to fix these fields!

### CODE REVIEW

**Comment by JOJ0** [comment]
This is so great @snejus. Finally, someone! So far it looks like it's the right thing to do. I'll try find time to play around with it in the next few days and report back here.

**Comment by JOJ0** [comment]
Hi @snejus, finally found some time to ressurect my beets-dev environment to play around a little. Unfortunately not the best results yet. It could certainly be that my dev-library is totally messed up. I didn't thoroughly check or clean it up yet. This is what I get:

### Tests run fine

```
pytest -v test/test_autotag.py -k test_ensure_consistent
=================================================================================== test session starts ===================================================================================
platform linux -- Python 3.11.3, pytest-7.4.0, pluggy-1.2.0 -- /home/jojo/.pyenv/versions/beets/bin/python
cachedir: /tmp/pytest_cache
rootdir: /home/jojo/git/beets
configfile: setup.cfg
plugins: anyio-3.7.0
collected 105 items / 84 deselected / 21 selected

test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value0-mb_artistid-mb_artistids] PASSED                                                                         [ 1/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value0-mb_albumartistid-mb_albumartistids] PASSED                                                               [ 2/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value0-albumtype-albumtypes] PASSED                                                                             [ 3/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value1-mb_artistid-mb_artistids] PASSED                                                                         [ 4/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value1-mb_albumartistid-mb_albumartistids] PASSED                                                               [ 5/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value1-albumtype-albumtypes] PASSED                                                                             [ 6/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value2-mb_artistid-mb_artistids] PASSED                                                                         [ 7/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value2-mb_albumartistid-mb_albumartistids] PASSED                                                               [ 8/21]
test/test_autotag.py::test_ensure_consistent_list_fields[None-list_value2-albumtype-albumtypes] PASSED                                                                             [ 9/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value3-mb_artistid-mb_artistids] PASSED                                                                            [10/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value3-mb_albumartistid-mb_albumartistids] PASSED                                                                  [11/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value3-albumtype-albumtypes] PASSED                                                                                [12/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value4-mb_artistid-mb_artistids] PASSED                                                                            [13/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value4-mb_albumartistid-mb_albumartistids] PASSED                                                                  [14/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value4-albumtype-albumtypes] PASSED                                                                                [15/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value5-mb_artistid-mb_artistids] PASSED                                                                            [16/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value5-mb_albumartistid-mb_albumartistids] PASSED                                                                  [17/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value5-albumtype-albumtypes] PASSED                                                                                [18/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value6-mb_artistid-mb_artistids] PASSED                                                                            [19/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value6-mb_albumartistid-mb_albumartistids] PASSED                                                                  [20/21]
test/test_autotag.py::test_ensure_consistent_list_fields[1-list_value6-albumtype-albumtypes] PASSED                                                                                [21/21]

============================================================================ 21 passed, 84 deselected in 0.19s ============================================================================
```

### beet write still gives me loads of these

...
Nikki Nair - Renormalization Support Group - 03 - It’s NP-Complicated  (2022/Unknown)  [APHA027] f:MP3-128kCBR a:70-01/ASIS |Electronic, Breakbeat, Electro
  mb_artistid:  -> 6341383
  mb_albumartistid:  -> 6341383
  albumtype:  -> 12"
Sending event: write
Sending event: after_write
Sending event: database_change
Nikki Nair - Renormalization Support Group - 04 - Donut Time (2022/Unknown)  [APHA027] f:MP3-128kCBR a:70-01/ASIS |Electronic, Breakbeat, Electro
  mb_artistid:  -> 6341383
  mb_albumartistid:  -> 6341383
  albumtype:  -> 12"
Sending event: write
Sending event: after_write
Sending event: database_change
Sending event: cli_exit
smartplaylist: Updating 21 smart playlists...
...

**Comment by snejus** [comment]
I forgot to mention - you need to reimport your music first! Try `beet import -LI` some of these albums, and then re-run `beet write`?

**Comment by snejus** [comment]
`beet mbsync` may also do the job and it should be much quicker

**Comment by JOJ0** [comment]
Thanks for the hint! Yeah mbsync is a good idea but reimport for one album was good enough for now. Much better now:

```
Nikki Nair - Renormalization Support Group - 01 - Plug  (2022/UK)  [APHA027] f:MP3-128kCBR a:70-01/ASIS |Electronic, Breakbeat, Electro
  albumtype:  -> 12", 33 ⅓ RPM, EP
Nikki Nair - Renormalization Support Group - 02 - Where Are U  (2022/UK)  [APHA027] f:MP3-128kCBR a:70-01/ASIS |Electronic, Breakbeat, Electro
  albumtype:  -> 12", 33 ⅓ RPM, EP
Nikki Nair - Renormalization Support Group - 03 - It’s NP-Complicated  (2022/UK)  [APHA027] f:MP3-128kCBR a:70-01/ASIS |Electronic, Breakbeat, Electro
  albumtype:  -> 12", 33 ⅓ RPM, EP
Nikki Nair - Renormalization Support Group - 04 - Donut Time (2022/UK)  [APHA027] f:MP3-128kCBR a:70-01/ASIS |Electronic, Breakbeat, Electro
  albumtype:  -> 12", 33 ⅓ RPM, EP
  ```

but note that `data_source` is Discogs in this case! Maybe we have to fix something in the Discogs plugin? Or it hast to do with the weird characters in there?


data in db is now:

```
$ sqlite3 ~/.config/devbeets/library.db "select artist,title,mb_artistid,mb_albumartistid,albumtype from items where mb_artistid == '6341383';"
Nikki Nair|Plug |6341383|6341383|12", 33 ⅓ RPM, EP
Nikki Nair|Where Are U |6341383|6341383|12", 33 ⅓ RPM, EP
Nikki Nair|It’s NP-Complicated |6341383|6341383|12", 33 ⅓ RPM, EP
Nikki Nair|Donut Time|6341383|6341383|12", 33 ⅓ RPM, EP
```

so at least this albumtype data _looks_ properly

**Comment by JOJ0** [comment]
Did an mbsync and all MusicBrainz tagged items are fine now btw! 👍

File: test/test_autotag.py
Comment (JOJ0) [inline test/test_autotag.py:1057]: Not savvy with pytest yet. Can you explain in 3 words what this does?

File: test/test_autotag.py
Comment (JOJ0) [inline test/test_autotag.py:1069]: and this!

Comment (JOJ0) [reply test/test_autotag.py:1069]: I'd like to augment this data with the data I find for albumtype in my (possibly broken) dev-library, but I need to really understand it first. Thanks!

File: test/test_autotag.py
Comment (snejus) [reply test/test_autotag.py:1057]: These are test cases: it tests each `(single_field, list_field)` combination against each `(single_value, list_value)` combination.

File: test/test_autotag.py
Comment (JOJ0) [reply test/test_autotag.py:1069]: Well I'm not 100% sure what I'm doing but I think I understood roughly what this is doing.

I tried this with mouse-copy/pasted data from my terminal output.

```
@pytest.mark.parametrize(
    "single_value,list_value",
    [
        (None, []),
        (None, ["1"]),
        (None, ["1", "2"]),
        ("1", []),
        ("1", ["1"]),
        ("1", ["1", "2"]),
        ("1", ["2", "1"]),
        ('12", 33 ⅓ RPM, EP', ['1', '12", 33 ⅓ RPM, EP']),
        ('12", 33 ⅓ RPM, EP', ['12", 33 ⅓ RPM, EP', 2]),
    ],
)
```

```
test/test_autotag.py::test_ensure_consistent_list_fields[12", 33 \u2153 RPM, EP-list_value7-mb_artistid-mb_artistids] PASSED                                                       [22/27]
test/test_autotag.py::test_ensure_consistent_list_fields[12", 33 \u2153 RPM, EP-list_value7-mb_albumartistid-mb_albumartistids] PASSED                                             [23/27]
test/test_autotag.py::test_ensure_consistent_list_fields[12", 33 \u2153 RPM, EP-list_value7-albumtype-albumtypes] PASSED                                                           [24/27]
test/test_autotag.py::test_ensure_consistent_list_fields[12", 33 \u2153 RPM, EP-list_value8-mb_artistid-mb_artistids] PASSED                                                       [25/27]
test/test_autotag.py::test_ensure_consistent_list_fields[12", 33 \u2153 RPM, EP-list_value8-mb_albumartistid-mb_albumartistids] PASSED                                             [26/27]
test/test_autotag.py::test_ensure_consistent_list_fields[12", 33 \u2153 RPM, EP-list_value8-albumtype-albumtypes] PASSED                                                           [27/27]
```

Comment (snejus) [reply test/test_autotag.py:1069]: That's the test cases for `single_field` (like `mb_artistid`) and `list_field` (like `mb_artistids`), covering cases where `mb_artistid` is not set, or `mb_artistids` is missing, or where first item in `mb_artistids` is not equal to `mb_artistid`.

Below it asserts that either both `mb_artistid` and `mb_artistids` are empty, or `mb_artistid` == `mb_artistids[0]` once the corrections are made by `ensure_consistent_list_fields`.

This test runs for each of the field pairs (`mb_albumartistid` / `mb_albumartistids` and `albumtype` / `albumtypes`).

**Comment by JOJ0** [comment]
I have a suspicion why albumtype is a problem with Discogs plugin. It only saves a single field `albumtype`, but not `albumtypes`.

**Comment by snejus** [comment]
> I have a suspicion why albumtype is a problem with Discogs plugin. It only saves a single field `albumtype`, but not `albumtypes`.

This change should handle such cases and set `albumtypes = [albumtype]`

**Comment by JOJ0** [comment]
> > I have a suspicion why albumtype is a problem with Discogs plugin. It only saves a single field `albumtype`, but not `albumtypes`.
> 
> This change should handle such cases and set `albumtypes = [albumtype]`

Ok I understand, that's odd. Anyway I just scrambled together some code that really sets albumtypes in the first place on a re/import. Do you mind if I push a commit to this branch? Just so we can talk about it better. You can revert it later on.

**Comment by snejus** [comment]
go ahead!

File: test/test_autotag.py
Comment (JOJ0) [reply test/test_autotag.py:1069]: Ah ok so now I understand that as you mentioned below, actually an empty albumtypes is tested here, as we have it in the Discogs plugin. This one I guess:

```
        ("1", []),
```

But wait, the Discogs plugin does not have albumtypes as a list _at all_, what is actually happening if you ask beets to give Item.nonexistent? Will it return None?

Should we add this to the test?

```
        ("1", None),
```

Comment (snejus) [reply test/test_autotag.py:1069]: See `AlbumInfo.__init__` implementation: if `albumtypes` is not given it's set to an empty list.

File: beetsplug/discogs.py
Comment (snejus) [inline beetsplug/discogs.py:437]: You'd want to set `albumtypes = []` for clarity here.

Actually `discogs` plugin has been updated heavily on my end, including calculation of the `albumtypes` field. Let me try to dig this out

File: beetsplug/discogs.py
Comment (JOJ0) [inline beetsplug/discogs.py:442]: Sorry that is clumsy code and I didn't look into the helper function someone provided (`get_media_and_albumtype()`) yet, but you get the point. Now a list field `albumtypes` is existing and yes I make it easy for `albumtype` being fixed here already. 

Actually setting albumtypes to an empty list here would have been sufficient, right?

File: beetsplug/discogs.py
Comment (snejus) [reply beetsplug/discogs.py:437]: Maybe we should leave this out for another PR, I'm thinking?

File: test/test_autotag.py
Comment (JOJ0) [reply test/test_autotag.py:1069]: Ok, damn it. So that's not it.

File: beetsplug/discogs.py
Comment (JOJ0) [reply beetsplug/discogs.py:442]: As you describe below. That's not it.... :-/

Comment (snejus) [reply beetsplug/discogs.py:442]: See my comment here: https://github.com/beetbox/beets/pull/5540#discussion_r1885384710. `albumtypes` _does_ get set to an empty list even when it's not provided

**Comment by snejus** [comment]
Could you by any chance provide `beet info` and `beet info -l` outputs for one of those tracks that still writes `albumtype` field?

File: beetsplug/discogs.py
Comment (JOJ0) [reply beetsplug/discogs.py:437]: Yeah exactly, had the same thought. I'm debugging around here already like crazy....but so far I think this PR is ready to merge.

Please reset or revert the branch to what it was before my commit and I think we should merge it now and note in an issue that there might something to fix with discogs plug.

Comment (snejus) [reply beetsplug/discogs.py:437]: Cool, will do this now!

Comment (snejus) [reply beetsplug/discogs.py]: Just need an approval and we're good to go for now.

File: test/test_autotag.py
Comment (snejus) [reply test/test_autotag.py:1069]: Hopefully your `beet info` outputs will give some insight into what needs to be adjusted in discogs!

**Comment by JOJ0** [comment]
There you go:

```
MORESOUNDS - Show Your Respect - 02 - Forward (2021/UK)  [APHALTD003] f:MP3-128kCBR a:70-01/ASIS |Electronic, Drum n Bass, Halftime
  mb_artistid:  -> 1253344
  mb_albumartistid:  -> 1253344
  albumtype:  -> 10", Limited Edition, White Label
```

```
$ beet info -l forward | grep albumtype
           albumtype: 10", Limited Edition, White Label
          albumtypes:
```



```
$ beet info forward | grep albumtype
```

(Missing entirely from the file tag.)

Sufficient or you really want to see all tags and lib items?


database, just in case:

```
sqlite3 ~/.config/devbeets/library.db "select artist,title,mb_artistid,albumtypes,albumtype from items where artist='Chimpo';"
Chimpo|Rig Doctor|645279||12"
Chimpo|Luv NRG|645279||12"
Chimpo|Buzz Army|645279||12"
Chimpo|Broad|645279||12"
```

File: beetsplug/discogs.py
Comment (JOJ0) [reply beetsplug/discogs.py]: One thing: Were would be documented that users need to reimport? Maybe a hint in the changelog? Something like: Read PR description how to fix?

**Comment by snejus** [comment]
Thanks. I messaged you on Zulip asking for one of the files in question so that I can test it out

File: beetsplug/discogs.py
Comment (snejus) [reply beetsplug/discogs.py]: Good point, I'll give this a mention in the changelog!

Comment (snejus) [reply beetsplug/discogs.py]: done 👍🏼

**Comment by snejus** [comment]
I found that album metadata was applied after these corrections, where `albumtypes` field was overwritten with an empty list (since `discogs` does not populate any). I made an adjustment where these corrections are applied _after_ album metadata was applied. Also added these  corrections to singletons.

File: beets/autotag/__init__.py
Comment (JOJ0) [inline beets/autotag/__init__.py:151]: Ok, so these hasattr checks exist to make this function work for Album model too. Got it.

File: beets/autotag/__init__.py
Comment (JOJ0) [inline beets/autotag/__init__.py:146]: BTW: Thanks for reminding me what the walrus operator does again! ;-) I should use it more often. In this specific case it saves quite a lot of lines!

File: beets/autotag/__init__.py
Comment (JOJ0) [inline beets/autotag/__init__.py:153]: It took me some time to understand this oneliner, but it actually was a lack of Python knowledge,. I wasn't aware of what `dict.fromkeys()` is doing. I was just trying to solve this using a dictionary comprehension but in the end, it does not get more readable:

```
m.mb_artistids = list({key: None for key in [aid, *m.mb_artistids]})
```

So I guess your solution is the most concise and best readable. Just brainstorming further, the default for all values dict.fromkeys() sets is `None`. One must know that, so when we make this explicit it would look like this:

```
m.mb_artistids = list(dict.fromkeys([aid, *m.mb_artistids], None))
```

```
m.mb_artistids = [
    dict.fromkeys([aid, *m.mb_artistids], None)
]
```

Since most of the time Python sets a default of None "for things" (eg. dict.get()) when it's not explicitly given, I guess that idea again does not make it more readable.

So yes, leave it as-is! Haha, just wanted to waste some lines of text speaking it out loud 😆

File: beets/autotag/__init__.py
Comment (JOJ0) [inline beets/autotag/__init__.py:192]: So this was required to fix the issue with Discogs tagged files that have `albumtypes` missing, right?

But nevertheless, I'll look into adding an `albumtypes` list field to the Discogs plugin. It just makes more sense to store those multiple values we usually get from Discogs accordingly. Haven't looked into your latest plugin changes though! Will do!

File: beets/autotag/__init__.py
Comment (JOJ0) [inline beets/autotag/__init__.py:183]: This is for singletons?

**Review by JOJ0 — APPROVED** [review]
Let's merge it! 🎉 

I still found ways to trigger the issue but that was with files that I imported as-is, so manual corrections using `beet mod albumtype=something query` and `beet mod albumtypes=something query` are the way to fix those. The `apply_metadata...` methods are not run in that case, thus it's not possible to correct those on (re)import. We shouldn't try to fix that here, since there "just is" quite some things in beets that depend on properly tagging with a tagging source. Importing as-is, just has flaws. It is what it is, and fixing that is for another day. :-)

Most of all: Thank you! Fixing this was long due. You secured yourself an entry in the books of "beets history" with that! ❤️

File: beets/autotag/__init__.py
Comment (JOJ0) [reply beets/autotag/__init__.py:153]: btw, that last example does not work anyway! forget it!

Comment (snejus) [reply beets/autotag/__init__.py]: To be fair there's a lot going on in this one liner - I feel like it would be useful to find a small comment here with a summary.

See the last commit - I defined `unique_list` as a separate util function and deduped the logic that sets the fields.

File: beets/autotag/__init__.py
Comment (snejus) [reply beets/autotag/__init__.py:183]: Indeed

File: beets/autotag/__init__.py
Comment (JOJ0) [reply beets/autotag/__init__.py]: Oh wow! Yeah that reads like a book now! More words but they are english! I like it! 

Some fun fact, while we are at it. I'm not entirely sure but I recently found a helper function in the last genre plugin: 

https://github.com/beetbox/beets/blob/22163d70a77449d83670e60ad3758474463de31b/beetsplug/lastgenre/__init__.py#L48-L51

Isn't that something very similar but using `set()` to achieve the "uniqueness". Set is unordered so the comprehension takes care of that. This is off-topic here but since I plan on finally finishing some lastgenre plugin things I might make use of your utility function instead of this one. What's the difference? Thanks for the help understanding it entirely beforehand! :-)

Comment (snejus) [reply beets/autotag/__init__.py]: I suspect this function originated before 3.7 was released (before then, `dict` keys were not ordered), so `list(dict.fromkeys(...))` wasn't a thing back then. The only builtin help that was available was `list(set(...))` however this is does not keep the original order.

Generally `deduplicate` implementation is a bit of a hack and (at least personally) requires a fair bit of mental effort to read. `list(dict.fromkeys(...))` is less verbose, [more than twice as fast](https://www.peterbe.com/plog/fastest-way-to-uniquify-a-list-in-python-3.6) and more well-known to people.

Comment (JOJ0) [reply beets/autotag/__init__.py]: Thank you so much for shedding some light on this. So helpful! Great! :-*

**Comment by arsaboo** [comment]
@snejus Seeing this error with `beet mbsync`:

```yaml
Traceback (most recent call last):
  File "/home/arsaboo/.local/bin/beet", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/arsaboo/.local/lib/python3.12/site-packages/beets/ui/__init__.py", line 1870, in main
    _raw_main(args)
  File "/home/arsaboo/.local/lib/python3.12/site-packages/beets/ui/__init__.py", line 1849, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/home/arsaboo/.local/lib/python3.12/site-packages/beetsplug/mbsync.py", line 73, in func
    self.albums(lib, query, move, pretend, write)
  File "/home/arsaboo/.local/lib/python3.12/site-packages/beetsplug/mbsync.py", line 183, in albums
    any_changed_item = items[0]
                       ~~~~~^^^
IndexError: list index out of range
```

I will create a PR to check for empty albums.

**Comment by snejus** [comment]
@arsaboo could you check whether this errors out before this change was merged? Doesn't seem like it's related

**Comment by arsaboo** [comment]
It is quite possible that those errors are unrelated to this PR (I ran `mbsync` for the first time today). May be it was another bug that was always there :)

**Comment by shreyasminocha** [comment]
Excited to see this in a release! Also, I think this supersedes #5075.

**Pull request merged**
